torchao/quantization/qat/api.py

```diff
@@ -239,6 +239,52 @@ def __setattr__(self, name: str, value: Any):
             super().__setattr__(name, value)
 
 
+def intx_quantization_aware_training(
+    activation_config: Optional[FakeQuantizeConfig] = None,
+    weight_config: Optional[FakeQuantizeConfig] = None,
+) -> torch.nn.Module:
+    """
+    Return a function that applies fake quantization to a `torch.nn.Module`.
+    to be used with the `quantize_` API.
+
+    Example usage::
+
+        from torchao.quantization import quantize_
+        from torchao.quantization.qat import FakeQuantizeConfig
+        activation_config = FakeQuantizeConfig(
+            torch.int8, "per_token", is_symmetric=False,
+        )
+        weight_config = FakeQuantizeConfig(
+            torch.int4, group_size=32, is_symmetric=True,
+        )
+        quantize_(model, intx_quantization_aware_training(activation_config, weight_config))
+    """
+
+    def _insert_fake_quantize(mod: torch.nn.Module):
+        """
+        Swap the given module with its corresponding fake quantized version.
+        """
+        from .embedding import FakeQuantizedEmbedding
+        from .linear import FakeQuantizedLinear
+
+        if isinstance(mod, torch.nn.Linear):
+            return FakeQuantizedLinear.from_linear(
+                mod,
+                activation_config,
+                weight_config,
+            )
+        elif isinstance(mod, torch.nn.Embedding):
+            if activation_config is not None:
+                raise ValueError("Embedding does not support QAT for activations")
+            return FakeQuantizedEmbedding.from_embedding(mod, weight_config)
```

Comment:
the function takes both `activation_config` and `weight_config`, but in this case we only use `weight_config`

should we be separating these two branches to 2 APIs?

Reply:
I feel it's better to only have one API, otherwise we'll need a separate API for each type of layer we support. Also it's more consistent with PTQ where we only have one `int8_weight_only` and let the `filter_fn` decide which layers to apply the transformation on (instead of having separate `int8_weight_only_linear` and `int8_weight_only_embedding`)

Reply:
oh the separation I'm referring to is to separate int8_act_int4_weight and int8_weight_only here, not linear v.s. embedding

Reply:
oh I see, you mean separate this into the following?
```
intx_activation_intx_weight_quantization_aware_training
intx_weight_only_quantization_aware_training
```
I can do that, but the naming seems a bit long... any suggestions?

Reply:
alternatively:
```
intx_quantization_aware_training (for both act + weight)
intx_weight_only_quantization_aware_training
```

Reply:
Ok, separated into these two for now. Please let me know if you have better suggestions for the naming:
```
intx_quantization_aware_training (for both act + weight)
intx_weight_only_quantization_aware_training
```

Reply:
maybe intx_quantization_aware_training can be `intx_dynamic_quantization_aware_training`?

Reply:
or just expand to match ptq and then add a `qat` as suffix, that might be clearer?